runtime/serde/serde-json/common/src/aws/smithy/kotlin/runtime/serde/json/JsonLexer.kt

```diff
@@ -0,0 +1,433 @@
+     */
+    private fun readNumber(): JsonToken {
+        val value = buildString {
+            if (peekChar() == '-') {
+                append(nextOrThrow())
+            }
+            readDigits(this)
+            if (peekChar() == '.') {
+                append(nextOrThrow())
+                readDigits(this)
+            }
+            if (peekChar() in EXP) {
+                append(nextOrThrow())
+                if (peekChar() in PLUS_MINUS) {
+                    append(nextOrThrow())
+                }
+                readDigits(this)
+            }
+        }
+        lexerCheck(value.isNotEmpty()) { "Invalid number, expected '-' || 0..9, found ${peekChar()}" }
+        return JsonToken.Number(value)
+    }
+
+    private fun readDigits(appendable: Appendable) {
+        while (peekChar() in DIGITS) {
+            appendable.append(nextOrThrow())
+        }
+    }
+
+    /**
+     * Read a quoted JSON string out of the stream
+     */
+    private fun readQuoted(): String {
+        consume('"')
+        // read bytes until a non-escaped end-quote
+        val start = idx
+        var chr = peekOrThrow()
+        var needsUnescaped = false
+        while (chr != '"') {
+            // handle escapes
+            when (chr) {
+                '\\' -> {
+                    needsUnescaped = true
+                    // consume escape backslash
+                    nextOrThrow()
+                    when (val byte = nextOrThrow()) {
+                        'u' -> {
+                            if (idx + 4 >= data.size) throw DeserializationException("Unexpected EOF")
+                            idx += 4
+                        }
+                        '\\', '/', '"', 'b', 'f', 'r', 'n', 't' -> { } // already consumed
+                        else -> fail("Invalid escape character: `$byte`", idx - 1)
+                    }
+                }
+                else -> {
+                    if (chr.isControl()) fail("Unexpected control character: `$chr`")
+                    idx++
+                }
+            }
+
+            chr = peekOrThrow()
+        }
+
+        val value = data.decodeToString(start, idx)
+        consume('"')
+        return if (needsUnescaped) value.unescape() else value
+    }
+
+    private fun readKeyword(): JsonToken = when (val ch = peekOrThrow()) {
+        't' -> readLiteral("true", JsonToken.Bool(true))
+        'f' -> readLiteral("false", JsonToken.Bool(false))
+        'n' -> readLiteral("null", JsonToken.Null)
+        else -> fail("Unable to handle keyword starting with '$ch'")
+    }
+
+    private fun readLiteral(expectedString: String, token: JsonToken): JsonToken {
+        consume(expectedString)
+        return token
+    }
+
+    /**
+     * Advance the cursor until next non-whitespace character is encountered
+     * @param peek Flag indicating if the next non-whitespace character should be consumed or peeked
+     */
+    private fun nextNonWhitespace(peek: Boolean = false): Char? {
```

Comment:
suggestion

I don't find any place where this function is called where `peek` is `false`, maybe unneeded.